src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -100,6 +114,144 @@ pub struct RSAggregateResult {
     pub type_mask: RSResultTypeMask,
 }
 
+impl RSAggregateResult {
+    /// Create a new aggregate result with the given capacity
+    pub fn new(cap: usize) -> Self {
+        Self {
+            num_children: 0,
+            children_cap: cap as c_int,
+            children: Self::new_children(cap),
+            type_mask: RSResultTypeMask::empty(),
+        }
+    }
+
+    /// Add a new child record to the aggregate result.
+    ///
+    /// # Undefined behaviour
+    /// The given `result` has to stay valid for the lifetime of this aggregate result. Else reading
+    /// from this result will cause undefined behaviour.
+    pub fn push(&mut self, result: &RSIndexResult) {
+        if self.num_children >= self.children_cap {
+            self.grow()
+        }
+
+        // SAFETY: `num_children` is within the valid range of the `children` pointer, else we would
+        // have grown above to make it valid.
+        let slot = unsafe { self.children.add(self.num_children as _) };
+
+        // SAFETY: we just used `add` to ensure `slot` is a valid pointer and is "aligned correctly"
+        unsafe {
+            *slot = result as *const _ as *mut _;
+        }
+
+        self.num_children += 1;
+        self.type_mask |= result.result_type;
+    }
+
+    /// Get a child record by index. Returns `None` if the index is out-of-bounds.
+    pub fn get(&self, index: usize) -> Option<&RSIndexResult> {
+        if index >= self.num_children as _ {
+            return None;
+        }
+
+        // SAFETY: `index` is less than `self.num_children` because of the check above. This means
+        // `index` will be within the valid range of the `children` pointer.
+        let child_ptr = unsafe { self.children.add(index) };
+
+        // SAFETY: we just used `add` to ensure `child_ptr` is a valid pointer
+        let child_ptr = unsafe { *child_ptr };
+
+        debug_assert!(
+            !child_ptr.is_null(),
+            "we should have checked correctly above"
+        );
+
+        // SAFETY: we expect the pointer to still be valid because of the condition placed on
+        // `push()` where we require the pointer to be valid for the lifetime of this
+        // aggregate result.
+        let child = unsafe { &*child_ptr };
+        Some(child)
+    }
+
+    fn grow(&mut self) {
+        let new_cap = if self.children_cap == 0 {
+            1
+        } else {
+            self.children_cap * 2
+        };
+
+        self.resize(new_cap)
+    }
+
+    fn resize(&mut self, new_cap: c_int) {
+        let new_children = Self::new_children(new_cap as _);
+
+        // Copy existing items to new array
+        if !self.children.is_null() && self.num_children > 0 {
```

Comment:
As in `get` you can check only `num_children` et assert the ptr is not NULL.

Reply:
Perfect, done in https://github.com/RediSearch/RediSearch/pull/6490/commits/ec4c5df0ecae3be1f276f6bda71b434abc20af9f